Sherlock/sherlock.h

```diff
@@ -87,208 +91,185 @@ constexpr const char* kDefaultTopLevelName = "TopLevelTransaction";
-        own_data_(schema_namespace_name_),
-        publisher_(std::make_unique<publisher_t>(own_data_)),
-        authority_(StreamDataAuthority::Own) {}
-
-  StreamImpl(const ss::StreamNamespaceName& namespace_name)
-      : schema_namespace_name_(namespace_name),
-        schema_as_object_(StaticConstructSchemaAsObject(schema_namespace_name_)),
-        own_data_(schema_namespace_name_),
-        publisher_(std::make_unique<publisher_t>(own_data_)),
-        authority_(StreamDataAuthority::Own) {}
+ private:
+  // Magic to enable `current::MakeOwned<Stream>` create instances of `Stream`.
+  struct PrivateConstructorWithoutNamespace {};
+  struct PrivateConstructorWithNamespace {};
+  friend struct sync::impl::UniqueInstance<Stream>;
 
-  template <typename X, typename... XS, class = std::enable_if_t<!std::is_same<X, ss::StreamNamespaceName>::value>>
-  StreamImpl(X&& x, XS&&... xs)
+  template <typename... ARGS>
+  Stream(PrivateConstructorWithoutNamespace, ARGS&&... args)
       : schema_as_object_(StaticConstructSchemaAsObject(schema_namespace_name_)),
-        own_data_(schema_namespace_name_, std::forward<X>(x), std::forward<XS>(xs)...),
-        publisher_(std::make_unique<publisher_t>(own_data_)),
-        authority_(StreamDataAuthority::Own) {}
+        impl_(MakeOwned<impl_t>(schema_namespace_name_, std::forward<ARGS>(args)...)),
+        owned_publisher_(MakeOwned<publisher_t>(impl_)),
+        borrowable_publisher_(Value(owned_publisher_)) {}
 
-  template <typename X, typename... XS>
-  StreamImpl(const ss::StreamNamespaceName& namespace_name, X&& x, XS&&... xs)
+  template <typename... ARGS>
+  Stream(PrivateConstructorWithNamespace, const ss::StreamNamespaceName& namespace_name, ARGS&&... args)
       : schema_namespace_name_(namespace_name),
         schema_as_object_(StaticConstructSchemaAsObject(schema_namespace_name_)),
-        own_data_(schema_namespace_name_, std::forward<X>(x), std::forward<XS>(xs)...),
-        publisher_(std::make_unique<publisher_t>(own_data_)),
-        authority_(StreamDataAuthority::Own) {}
+        impl_(MakeOwned<impl_t>(schema_namespace_name_, std::forward<ARGS>(args)...)),
+        owned_publisher_(MakeOwned<publisher_t>(impl_)),
+        borrowable_publisher_(Value(owned_publisher_)) {}
 
-  StreamImpl(StreamImpl&& rhs)
-      : schema_namespace_name_(rhs.schema_namespace_name_),
-        schema_as_object_(StaticConstructSchemaAsObject(schema_namespace_name_)),
-        own_data_(std::move(rhs.own_data_)),
-        publisher_(std::move(rhs.publisher_)),
-        authority_(rhs.authority_) {
-    rhs.authority_ = StreamDataAuthority::External;
+ public:
+  // `Publisher()`: The caller assumes full responsibility for making sure the underlying stream
+  // is not destroyed while it is being published to.
+  // CAN THROW `PublisherNotAvailableException`.
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
```

Comment:
I have a gut feeling that exposing `template <current::locks::MutexLockStatus MLS>` functions to the end user is not good :(